Perform topological sort to order components

cmacmackin opened the pull request:
Using the access control information introduced in #421, this PR makes it possible for Hermes-3 to work out the order of components at run-time. This will make things far simpler and more robust for users. It will also fail faster if there is an unsatisfiable or circular dependency.

Closes #384.

cmacmackin committed:
Added unit test for topological sorting of components

cmacmackin committed:
Ensure ordering of components accounts for pre-set state variables

cmacmackin committed:
Added support for sorting components based on section permissions

cmacmackin commented:
This is nearly done. Remaining tasks:

- [ ] Merge #421 and rebase.
- [ ] There is one integration test giving me strange h5py errors on my computer. It doesn't look to be related to any changes I've made. We'll see what happens in CI. I seem to recall we saw something similar elsewhere at one point.
- [ ] Update documentation.

cmacmackin committed:
Mention new sorting and remove references to component ordering

cmacmackin committed:
Check results are the same regardless of component order

cmacmackin committed:
Check only one component can make final write

cmacmackin committed:
Add better comments explaining component sorting process

cmacmackin committed:
Responded to some comments by @ZedThree

The comments were in PR #445, but some of them actually relate to work
done in #428. Only the latter are dealt with in this PR.

ZedThree commented on src/component_scheduler.cxx:
```diff
@@ -1,13 +1,314 @@
+
+const std::set<std::string> ComponentScheduler::predeclared_variables = {
+    "time",          "linear",      "units:inv_meters_cubed", "units:eV", "units:Tesla",
+    "units:seconds", "units:meters"};
```
Given this is only used in this file, we can move the declaration out of the header and just have it here.

cmacmackin replied:
Done.

mikekryjak commented on docs/sphinx/developer.rst:
```diff
@@ -662,8 +679,10 @@ of components
    # options to control component3
 
 This will create three components, which will be run in the order
```
I think this is incorrect, because the run order will now be determined by the component scheduler.

mikekryjak commented on include/permissions.hxx:
```diff
@@ -308,6 +317,20 @@ inline Permissions::VarRights writeBoundary(std::string varname) {
+/// permissions on the boundaries and read permissions for the
+/// interior.
+///
+/// FIXME: Currently these permissiosn are not expressed properly, due
```
is this FIXME still active? what does it take to fix it and what are the possible risks of leaving it unfixed?

mikekryjak commented:
I missed this in #421, but I see that you added a long-needed feature - a better way to determine species type which is based on the charge:
https://github.com/boutproject/hermes-3/blob/8fc17e0cad037999b97a31cdf53d6558cdd859d8/src/component_scheduler.cxx#L339-L352

There is already a function that does the same in `hermes_utils` and is used in a few places in the code:
https://github.com/boutproject/hermes-3/blob/053c9bb70b66ffd3ac603477f76c5866a0b9e33c/include/hermes_utils.hxx#L61-L70

Ideally there should be only one tool to do this, and your new one seems more robust. Is there any reason not to replace the one in `hermes_utils` with the new one?

cmacmackin commented:
There are some places (e.g., when writing tests) where it was convenient to just be able to have a list of species names and use the old heuristics to categorise them. Probably not a good enough reason to keep that though.

ZedThree reviewed (approved):
LGTM, thanks @cmacmackin!

There's some trivial bits that I'm happy to fix myself

ZedThree commented on src/component_scheduler.cxx:
```diff
@@ -1,13 +1,314 @@
+/// In pratice, `item` here represents the index of a particular
+/// component. The indices of the dependencies of `item` are stored in
+/// the corresponding element of `dependencies`.
+void topological_sort(const std::vector<std::set<size_t>>& dependencies, size_t item,
```
It's a guard against One Definition Rule (ODR) violations -- basically ensures that functions have names unique to the translation unit, and so if there happens to be another function with the same name somewhere else, they won't clash at link time.

cmacmackin replied:
Done this now.

cmacmackin committed:
Switch transform method to using GuardedOptions

Note that components still aren't setting up any permissions, so there
would be runtime errors when executing the transform
methods. Furthermore, there are some missing methods I realise I need
for GuardedOptions. The unit tests are also failing to compile,
although I don't understand what the problem is for them, yet.

cmacmackin committed:
Applied linter suggestions

cmacmackin committed:
Print meaningful error messages for circular dependencies

cmacmackin committed:
Addressed various comments on PR

cmacmackin commented on docs/sphinx/developer.rst:
```diff
@@ -577,6 +577,17 @@ The `name` is a string labelling the instance. The `alloptions` tree contains at
+body of the constructor of your component, based on what
+configurations were specified in ``alloptions``. You should give read
+and write permissions to the minimum number of variables necessary, to
+avoid circular dependencies arising among components.
```
Fixed

cmacmackin commented on src/component_scheduler.cxx:
```diff
@@ -1,13 +1,314 @@
+  for (const auto& name : conditional_names) {
+    // FIXME: this isn't an empty set
+    //result.insert({name, {name}});
+    result.insert({name, {}});
```
Deleted this

cmacmackin commented on src/component_scheduler.cxx:
```diff
@@ -1,13 +1,314 @@
+
+/// Get all the parent sections of a variable "path". Sections are
+/// separated by colons in the path.
+std::set<std::string> getParents(const std::string& name) {
```
I've expanded the docstring to explain this better.

bendudson committed:
external_apar: Use NamedComponent

Use CRTP to get component name for registration.

cmacmackin committed:
Improve documentation/description for component sorting algorithm

cmacmackin committed:
Add ability to order components automatically

cmacmackin committed:
Fixed some incorrect access controls causing errors

cmacmackin commented:
There are a few examples that failed when I applied the topological sort to their components. 

In `other/linear/annulus-isothermal-he-emag`, both `electromagnetic` and `sheath_boundary` attempt to write the boundaries of the momentum and velocity of electrons and helium ions. This is something both components legitimately do. I'm guessing that `sheath_boundary` is intended to overwrite the boundary values set by `electromagnetic`. I've updated the latter accordingly so it no longer sets boundary values for these variables.

`tokamak-1D/extra/1D-hydrogen-equalT` had both `sheath_boundary` and `transform` trying to write the boundaries of deuterium ion temperature. I updated the permissions of `transform` so it is only considered to read and write the interior of a variable, not the boundaries. This could hypothetically cause problems in the future if someone expects it to write the boundaries. I could potentially make it configurable whether it write the boundaries or not.

`tokamak-2D/heat-transport` had a small mistake which meant it was asking for permission to read temperatures in the boundaries, when it didn't need to. This mistake resulted in an runtime error due to a circular dependency between components. The circular dependency went away once the permissions were corrected.

I didn't have time to wait for all the tests to finish before I left for the day, so we'll have to see whether these changes introduced any regressions.
